gsk4/src/render_node.rs

```diff
@@ -0,0 +1,173 @@
+        unref => |ptr| ffi::gsk_render_node_unref(ptr),
+    }
+}
+
+impl StaticType for RenderNode {
+    #[doc(alias = "gsk_render_node_type_get_type")]
+    fn static_type() -> Type {
+        unsafe { from_glib(ffi::gsk_render_node_type_get_type()) }
+    }
+}
+
+impl RenderNode {
+    //#[doc(alias = "gsk_render_node_deserialize")]
+    //pub fn deserialize(bytes: &glib::Bytes, error_func: /*Unimplemented*/FnMut(/*Ignored*/ParseLocation, /*Ignored*/ParseLocation, &glib::Error), user_data: /*Unimplemented*/Option<Fundamental: Pointer>) -> Option<RenderNode> {
+    //    unsafe { TODO: call ffi:gsk_render_node_deserialize() }
+    //}
+
+    pub fn downcast<T: IsARenderNode>(self) -> Result<T, RenderNode> {
+        unsafe {
+            if self.get_node_type() == T::NODE_TYPE {
+                Ok(from_glib_full(self.to_glib_full()))
+            } else {
+                Err(self)
+            }
+        }
+    }
+
+    pub fn downcast_ref<T: IsARenderNode>(&self) -> Option<&T> {
+        unsafe {
+            if self.get_node_type() == T::NODE_TYPE {
+                Some(&*(self as *const RenderNode as *const T))
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn draw(&self, cr: &cairo::Context) {
+        unsafe {
+            ffi::gsk_render_node_draw(self.to_glib_none().0, mut_override(cr.to_glib_none().0));
+        }
+    }
+
+    pub fn get_bounds(&self) -> graphene::Rect {
+        unsafe {
+            let mut bounds = graphene::Rect::uninitialized();
+            ffi::gsk_render_node_get_bounds(self.to_glib_none().0, bounds.to_glib_none_mut().0);
+            bounds
+        }
+    }
+
+    pub fn get_node_type(&self) -> RenderNodeType {
+        unsafe { from_glib(ffi::gsk_render_node_get_node_type(self.to_glib_none().0)) }
+    }
+
+    pub fn serialize(&self) -> glib::Bytes {
+        unsafe { from_glib_full(ffi::gsk_render_node_serialize(self.to_glib_none().0)) }
+    }
+
+    pub fn write_to_file<P: AsRef<Path>>(&self, filename: P) -> Result<(), glib::Error> {
+        unsafe {
+            let mut error = ptr::null_mut();
+            let _ = ffi::gsk_render_node_write_to_file(
+                self.to_glib_none().0,
+                filename.as_ref().to_glib_none().0,
+                &mut error,
+            );
+            if error.is_null() {
+                Ok(())
+            } else {
+                Err(from_glib_full(error))
+            }
+        }
+    }
+}
+
+pub trait IsARenderNode:
+    StaticType
+    + FromGlibPtrFull<*mut ffi::GskRenderNode>
+    + std::convert::AsRef<crate::RenderNode>
+    + 'static
+{
+    const NODE_TYPE: RenderNodeType;
+    fn upcast(self) -> RenderNode;
+    fn upcast_ref(&self) -> &RenderNode;
+}
+
+pub const NONE_RENDER_NODE: Option<&RenderNode> = None;
+
+impl fmt::Display for RenderNode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("RenderNode")
+    }
+}
+
+#[macro_export]
```

Comment:
We probably don't want to export this outside the crate

Reply:
There might be something that I don't understand here, but I couldn't figure out how to make a crate-scoped macro.  It seems like I can either define a macro that's valid for the rest of the file, or I can export it.  Since I really wanted to keep these render nodes in their own separate files, I was forced to export it.

Reply:
https://danielkeep.github.io/tlborm/book/mbe-min-scoping.html  Ah, yes, I needed to do that, and also use `#[macro_use]` to allow the macro to escape the module.